boshio/boshio.go

```diff
@@ -0,0 +1,227 @@
+package boshio
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync"
+)
+
+type Stemcell struct {
+	Name    string
+	Version string
+	Light   *Metadata `json:"light"`
+	Regular *Metadata `json:"regular"`
+}
+
+type Metadata struct {
+	URL  string
+	Size int64
+	MD5  string
+	SHA1 string
+}
+
+type bar interface {
+	SetTotal(contentLength int64)
+	Add(totalWritten int) int
+	Kickoff()
+	Finish()
+}
+
+type ranger interface {
+	BuildRange(contentLength int64) ([]string, error)
+}
+
+func (s Stemcell) Details() Metadata {
+	if s.Light != nil {
+		return *s.Light
+	}
+
+	return *s.Regular
+}
+
+type Client struct {
+	Host                 string
+	bar                  bar
+	ranger               ranger
+	stemcellMetadataPath string
+	stemcellDownloadPath string
+}
+
+func NewClient(b bar, r ranger) *Client {
```

Comment:
These types should be exported IMO since they're used in an exported function. I've always found that to be a weird quirk of Go's design; it inhibits discoverability of the interface's definition.